src/ai/backend/common/runner/runner.py

```diff
@@ -0,0 +1,111 @@
+import asyncio
+from collections.abc import Sequence
+from typing import Generic, Optional, Protocol, TypeVar
+
+TResourceCtx = TypeVar("TResourceCtx")
+TResourceCtx_Co = TypeVar("TResourceCtx_Co", covariant=True)
+TResourceCtx_Contra = TypeVar("TResourceCtx_Contra", contravariant=True)
+
+
+class ResourceCtx(Protocol[TResourceCtx_Co]):
+    async def open(self) -> TResourceCtx_Co:
+        pass
+
+    async def close(self) -> None:
+        pass
+
+
+class NopResourceCtx:
+    def __init__(self) -> None:
+        pass
+
+    async def open(self) -> None:
+        return None
+
+    async def close(self) -> None:
+        return None
+
+
+class Probe(Protocol[TResourceCtx_Contra]):
+    async def probe(self, resource_ctx: TResourceCtx_Contra) -> None:
+        pass
+
+
+class HeartbeatService(Protocol[TResourceCtx_Contra]):
+    async def heartbeat(self, resource_ctx: TResourceCtx_Contra) -> None:
+        pass
+
+
+class ProbeRunner(Generic[TResourceCtx]):
```

Comment:
Remove `Probe` prefix. but `Runner` is so generic name. please rename appropriate name.